File: next/kmp/app/wasmBackend/src/wasmJsMain/kotlin/org/dweb_browser/wasm_backend/httpServer.kt
@file:JsModule("./http-server.mjs")

package org.dweb_browser.wasm_backend

import kotlin.js.Promise

external class ReqResCtx {
  val url: String
  fun setHeader(key: JsString, value: JsString)
  fun write(data: JsAny)
  fun end()
}

external fun createHttpServer(
  port: JsNumber = definedExternally,
  onReq: (ReqResCtx) -> Unit
): Promise<JsNumber>

File: next/kmp/app/jsBackend/src/wasmJsMain/resources/http-server.mjs
import http from "node:http";

export class ReqResCtx {
  #req;
  #res;
  constructor(req, res) {
    this.#req = req;
    this.#res = res;
  }
  get url() {
    return this.#req.url;
  }
  get headers() {
    return this.#req.headers;
  }
  setHeader(key, value) {
    this.#res.setHeader(key, value);
  }
  write(data) {
    this.#res.write(data);
  }
  end() {
    this.#res.end();
  }
}

export function createHttpServer(port = 0, onReq) {
  return new Promise((resolve, reject) => {
    const server = http
      .createServer((req, res) => {
        onReq(new ReqResCtx(req, res));
      })
      .on("error", reject)
      .on("close", () => {
        console.log("closed");
      })
      .listen({ port, host: "0.0.0.0" }, () => {
        const address = server.address();
        resolve(address.port);
        // console.log(address);
      });
  });
}
